### PR TITLE
Refactor prefetch

### PR DESCRIPTION
**Pull request by jacquesqiao**
sparse updated parameter can be prefetched from pserver when data come in, but not recv all parameter after one mini batch.

## TODO

1. add VarType for prefetch (int_name, out_name, table_name)
2. optimize parameter_prefetch.cc, add more unit test.
3. unify is_distributed and remote_prefetch.

### CODE REVIEW

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 1]
```diff
@@ -113,14 +116,22 @@ bool RequestPrefetchHandler::Handle(const std::string& varname,
...
-      (*prefetch_var_name_to_prepared_ctx_)[varname].get(), scope);
-
+  if (table_name.empty()) {
+    auto var_desc = program_->Block(0).FindVar(out_var_name);
+    InitializeVariable(*outvar, var_desc->GetType());
+    executor_->RunPreparedContext(
+        (*prefetch_var_name_to_prepared_ctx_)[varname].get(), scope);
+  } else {
+    (*outvar)->GetMutable<framework::LoDTensor>();
+    auto lookup_table_op =
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/request_handler_impl.cc:130]: can cache op instances later, using key = table_name, varname, out_var_name

File: paddle/fluid/operators/distributed/send_recv.proto.in [hunk 2]
```diff
@@ -80,6 +80,7 @@ message VariableMessage {
   // when profile switches from 1 to 2.
   int64 profile = 11;
   int64 trainer_id = 12;
+  string table_name = 13;
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/send_recv.proto.in:83]: is it possible to add a `VarType` than attributes? for prefetch, it should be a `vector<std:string>` of length 3.

File: paddle/fluid/operators/lookup_table_op.h [hunk 3]
```diff
@@ -23,6 +23,12 @@ limitations under the License. */
 #include "paddle/fluid/framework/selected_rows.h"
 #include "paddle/fluid/operators/math/blas.h"
 
+#ifdef PADDLE_WITH_DISTRIBUTE
+
```
Comment (typhoonzero) [inline paddle/fluid/operators/lookup_table_op.h]: can remove the redundant blank lines

File: paddle/fluid/operators/lookup_table_op.h [hunk 4]
```diff
@@ -41,44 +47,66 @@ class LookupTableKernel : public framework::OpKernel<T> {
...
+    auto id_name = context.Inputs("Ids").front();
+    auto out_name = context.Outputs("Out").front();
+
+    // for remote prefetch
+    auto epmap = context.Attr<std::vector<std::string>>("epmap");
+    auto height_sections = context.Attr<std::vector<int>>("height_sections");
+    auto table_names = context.Attr<std::vector<std::string>>("table_names");
+
+    if (!height_sections.empty()) {
+// if emap is not empty, then the parameter will be fetched from remote
```
Comment (typhoonzero) [inline paddle/fluid/operators/lookup_table_op.h]: emap => epmap

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -326,6 +327,8 @@ def embedding(input,
     """
 
     helper = LayerHelper('embedding', **locals())
+    if remote_prefetch:
```
Comment (typhoonzero) [inline python/paddle/fluid/layers/nn.py:332]: remote_prefetch and is_distributed are confusing, any way to remove one?

File: paddle/fluid/operators/distributed/request_handler_impl.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/request_handler_impl.cc:130]: Ok~this will be done in next pr.

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -287,7 +287,8 @@ def embedding(input,
               is_distributed=False,
               padding_idx=None,
               param_attr=None,
-              dtype='float32'):
+              dtype='float32',
+              remote_prefetch=False):
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: doc?

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -326,6 +327,8 @@ def embedding(input,
     """
 
     helper = LayerHelper('embedding', **locals())
+    if remote_prefetch:
+        assert is_sparse is True and is_distributed is False
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:333]: So strange...is_distributed must be false to remote prefetch...Is remote_prefetch distributed or not?

File: paddle/fluid/operators/lookup_table_op.h (hunk 4 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/lookup_table_op.h]: done

File: paddle/fluid/operators/lookup_table_op.h (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/lookup_table_op.h]: done

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (jacquesqiao) [reply python/paddle/fluid/layers/nn.py:332]: remote_prefetch is deleted and will be replaced by is_sparse in the future.

File: python/paddle/fluid/layers/nn.py (hunk 7 above)
Comment (jacquesqiao) [reply python/paddle/fluid/layers/nn.py:333]: remote_prefetch  is deleted from the interface.

File: paddle/fluid/operators/distributed/parameter_prefetch.cc [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    auto& ids = splited_ids[i];
+    if (!ids.empty()) {
+      auto* id_tensor_data = id_tensor->mutable_data<int64_t>(
+          framework::make_ddim({static_cast<int64_t>(ids.size()), 1}), place);
+      memcpy(id_tensor_data, ids.data(), sizeof(int64_t) * ids.size());
+    }
+  }
+}
+
+static void MergeMultipleVarsIntoOneBySection(
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/parameter_prefetch.cc:100]: use static to mark private function is kind of C style, do you plan to refactor this?

File: paddle/fluid/operators/distributed/send_recv.proto.in (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/send_recv.proto.in:83]: will optimize in the future pr.

File: paddle/fluid/operators/distributed/parameter_prefetch.cc (hunk 8 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/parameter_prefetch.cc:100]: Yes, now we use height_section to split the ids,  in the future will add more hash method

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**